### ISSUE TITLE
TabularNeuralNetModel taking unnecessarily long to train

### ISSUE THREAD

**Issue by thvasilo** [awslabs/autogluon#528]
Hello,

Thanks for the great package, really interesting stuff!

I ran a minimal example using a credit card fraud detection dataset, available [here](https://www.kaggle.com/mlg-ulb/creditcardfraud). Since this is a highly unbalanced dataset, I used `balanced_accuracy` as my metric.

One thing I noticed was that, compared to the other models, the `TabularNeuralNetModel` took ages to train, and essentially blocked the whole operation for no significant gain in accuracy. Here's my `fit_summary()` output:

```
*** Summary of fit() ***
Estimated performance of each model:
                         model  score_val  pred_time_val     fit_time  pred_time_val_marginal  fit_time_marginal  stack_level  can_infer
0           LightGBMClassifier      1.000       0.007321     1.080862                0.007321           1.080862            0       True
1      weighted_ensemble_k0_l1      1.000       0.008837     2.189336                0.001516           1.108474            1       True
2          NeuralNetClassifier      1.000       0.067239  2348.194962                0.067239        2348.194962            0       True
3   RandomForestClassifierEntr      1.000       0.109879    54.354113                0.109879          54.354113            0       True
4     ExtraTreesClassifierGini      1.000       0.109977     8.654530                0.109977           8.654530            0       True
5     ExtraTreesClassifierEntr      1.000       0.110057     7.852540                0.110057           7.852540            0       True
6   RandomForestClassifierGini      1.000       0.110687    62.168143                0.110687          62.168143            0       True
7     LightGBMClassifierCustom      0.875       0.007210     1.102000                0.007210           1.102000            0       True
8           CatboostClassifier      0.875       0.007485     0.508587                0.007485           0.508587            0       True
9     KNeighborsClassifierDist      0.750       0.115204     1.596709                0.115204           1.596709            0       True
10    KNeighborsClassifierUnif      0.500       0.115374     1.607259                0.115374           1.607259            0       True
Number of models trained: 11
Types of models trained:
{'LGBModel', 'CatboostModel', 'KNNModel', 'TabularNeuralNetModel', 'WeightedEnsembleModel', 'RFModel', 'XTModel'}
Bagging used: False 
Stack-ensembling used: False 
Hyperparameter-tuning used: False 
User-specified hyperparameters:
{'default': {'NN': [{}], 'GBM': [{}], 'CAT': [{}], 'RF': [{'criterion': 'gini', '_ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'entropy', '_ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}}], 'XT': [{'criterion': 'gini', '_ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'entropy', '_ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}}], 'KNN': [{'weights': 'uniform', '_ag_args': {'name_suffix': 'Unif'}}, {'weights': 'distance', '_ag_args': {'name_suffix': 'Dist'}}], 'custom': [{'num_boost_round': 10000, 'num_threads': -1, 'objective': 'binary', 'verbose': -1, 'boosting_type': 'gbdt', 'learning_rate': 0.03, 'num_leaves': 128, 'feature_fraction': 0.9, 'min_data_in_leaf': 5, 'two_round': True, 'seed_value': 0, '_ag_args': {'model_type': 'GBM', 'name_suffix': 'Custom', 'disable_in_hpo': True}}]}}
Plot summary of models saved to file: AutogluonModels/ag-20200619_213033/SummaryOfModels.html
*** End of fit() summary ***
```
I'm running on a CPU machine with 16 vCores (ml.c5.4xlarge) and the NN takes ~2400sec to train compared to ~1sec for the winning LightGBM algorithm.

Given that, I would suggest either removing the NN from the defaults for TabularData, or including a clear warning of its computational cost on a CPU-only machine.

**Comment by austinmw** [comment]
This comment probably won't address your overall issue, but if I recall correctly, I think I was told it's preferable to use M instances over C instances for these models, so comparing that to a run with a ml.m5.24xlarge might be worth a shot.

**Comment by Innixma** [comment]
One thing to check is if you have intel-mkl installed. If so, uninstalling it will dramatically speed-up the NN training.

Furthermore, the reason this particular NN took a long time is because it had a val_score of 1.0, and the NN keeps training if its val_score doesn't decrease over N iterations, up to a maximum of 500 iterations. If you increase the amount of data being trained with AutoGluon (to at least 2000 rows or so), you would see that the NN will actually take far less time to train because it uses less iterations. Whether the NN should keep training when it has a perfect val_score is a valid question, as theoretically it is better to keep training in this scenario, but may not be what most users actually want, as it increases training times in these cases significantly.

In general, the NN can often take the most time of the standard models, however removing it isn't the ideal solution, as it offers a large improvement to many problems. Also, NN isn't always the most time intensive model, CatBoost can also take a very long time depending on the problem. Do note that NN is nearly the last model to be trained, so is more of a luxury model than a necessity given time_limits argument.

If you care about how long a given training will take, specify time_limits argument and it will limit the time the NN will train for. I assume you ran this without specifying time_limits, which essentially means AutoGluon will keep training until every model is finished, without caring about time. It is recommended to always specify time_limits if you are prototyping/want a quick solution.

A future improvement I plan to make is to allow for model-specific and generic time_limits on a model basis, rather than an overall .fit basis. This would limit each model to only take, say, 240 seconds to train.

**Comment by Innixma** [comment]
#531 should address some of this concern. Once merged, you can specify a new argument to .fit():

```
# Ensure that no single model trains for longer than 60 seconds
task.fit(..., ag_args_fit={'max_time_limit': 60})
```

**Comment by thvasilo** [comment]
Thanks @Innixma seems like #531 would indeed help in cases such as these.

I'm using a SageMaker notebook to test this out, doing `conda list | grep mkl` on my env gives

```
blas                      1.0                         mkl
mkl                       2020.0                      166
mkl-service               2.3.0            py36he904b0f_0
mkl_fft                   1.0.15           py36ha843d7b_0
mkl_random                1.1.0            py36hd6b4f25_0
```

> One thing to check is if you have intel-mkl installed. If so, uninstalling it will dramatically speed-up the NN training.

This seems counter-intuitive, you're saying removing intel-mkl will _speed up_ training?

Thanks for pointing out leaderboard, good to see accuracy on my test set in the end. LightGBM did end up being the best with the test set as well in this occasion. Will be interesting to compare when you add XGBoost.

**Comment by Innixma** [comment]
Regarding mkl, yes, uninstalling it generally results in a significant speedup in NN training time. I don't know why at the current time, but I have observed it in my tests.
